In [2]:
bilgi_tabani=[
     "Python, 1991 yılında Guido van Rossum tarafından geliştirilen, okunabilirliği yüksek bir programlama dilidir.",
    "PyTorch, Facebook (Meta) tarafından geliştirilen, derin öğrenme projelerinde yaygın kullanılan bir kütüphanedir.",
    "CNN (Convolutional Neural Network), özellikle görüntü işleme problemlerinde kullanılan bir derin öğrenme mimarisidir.",
    "Transformer mimarisi, 2017 yılında 'Attention is All You Need' makalesiyle tanıtılmış ve LLM'lerin temelini oluşturmuştur.",
    "Pandas, Python'da veri analizi ve manipülasyonu için kullanılan popüler bir kütüphanedir.",
    "Titanic veri seti, makine öğrenmesi eğitiminde en çok kullanılan başlangıç veri setlerinden biridir."
]
for i, metin in enumerate(bilgi_tabani):
  print(f"{i}:{metin}")

0:Python, 1991 yılında Guido van Rossum tarafından geliştirilen, okunabilirliği yüksek bir programlama dilidir.
1:PyTorch, Facebook (Meta) tarafından geliştirilen, derin öğrenme projelerinde yaygın kullanılan bir kütüphanedir.
2:CNN (Convolutional Neural Network), özellikle görüntü işleme problemlerinde kullanılan bir derin öğrenme mimarisidir.
3:Transformer mimarisi, 2017 yılında 'Attention is All You Need' makalesiyle tanıtılmış ve LLM'lerin temelini oluşturmuştur.
4:Pandas, Python'da veri analizi ve manipülasyonu için kullanılan popüler bir kütüphanedir.
5:Titanic veri seti, makine öğrenmesi eğitiminde en çok kullanılan başlangıç veri setlerinden biridir.


enumerate(bilgi_tabani) fonkiyonu bizim metnimizi satır satır ayırmayı sağlıyor

In [3]:
!pip install sentence-transformers --quiet
from sentence_transformers import SentenceTransformer

model_embed = SentenceTransformer('all-MiniLM-L6-v2')
embeddings = model_embed.encode(bilgi_tabani)

print(f"embedding boyutu: {embeddings.shape} ")
print(f"\nİlk cümlenin embedding vektörü (ilk 10 sayı):\n{embeddings[0][:10]}")


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors: reconstructing file:   0%|          |  0.00B / 90.9MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

embedding boyutu: (6, 384) 

İlk cümlenin embedding vektörü (ilk 10 sayı):
[-0.07054368  0.04185832 -0.05596847  0.00213226 -0.06764742 -0.02688359
 -0.03179614  0.04499089 -0.0185852  -0.05086725]


.encode(...) fonksiyonu, verdiğimiz 6 cümlelik listeyi (bilgi_tabani) tek tek alıp, her birini bir sayı vektörüne çeviriyor. Yani her cümle artık "Python, 1991 yılında..." gibi bir yazı değil, [0.23, -0.51, 0.88, ...] gibi bir sayı dizisi haline geliyor





.shape, bize elimizdeki verinin boyutunu gösterecek. Muhtemelen (6, 384) gibi bir sonuç çıkacak: 6 → bizim 6 cümlemiz, 384 → her cümlenin kaç sayıdan oluşan bir vektörle temsil edildiği (bu, all-MiniLM-L6-v2 modelinin sabit bir özelliği).

In [4]:
soru = "PyTorch'u kim geliştirdi?"

soru_embedding = model_embed.encode([soru])

from sklearn.metrics.pairwise import cosine_similarity

benzerlikler = cosine_similarity(soru_embedding, embeddings)

print("Her cümleyle benzerlik skorları:")
for i, skor in enumerate(benzerlikler[0]):
    print(f"{i}: {skor:.4f} → {bilgi_tabani[i]}")

Her cümleyle benzerlik skorları:
0: 0.4966 → Python, 1991 yılında Guido van Rossum tarafından geliştirilen, okunabilirliği yüksek bir programlama dilidir.
1: 0.4414 → PyTorch, Facebook (Meta) tarafından geliştirilen, derin öğrenme projelerinde yaygın kullanılan bir kütüphanedir.
2: 0.2449 → CNN (Convolutional Neural Network), özellikle görüntü işleme problemlerinde kullanılan bir derin öğrenme mimarisidir.
3: 0.3336 → Transformer mimarisi, 2017 yılında 'Attention is All You Need' makalesiyle tanıtılmış ve LLM'lerin temelini oluşturmuştur.
4: 0.3755 → Pandas, Python'da veri analizi ve manipülasyonu için kullanılan popüler bir kütüphanedir.
5: 0.2982 → Titanic veri seti, makine öğrenmesi eğitiminde en çok kullanılan başlangıç veri setlerinden biridir.


In [5]:
en_yuksek_index = benzerlikler[0].argmax()
en_alakali_cumle = bilgi_tabani[en_yuksek_index]

print(f"Soru: {soru}")
print(f"En alakalı bulunan bilgi: {en_alakali_cumle}")
print(f"Benzerlik skoru: {benzerlikler[0][en_yuksek_index]:.4f}")

Soru: PyTorch'u kim geliştirdi?
En alakalı bulunan bilgi: Python, 1991 yılında Guido van Rossum tarafından geliştirilen, okunabilirliği yüksek bir programlama dilidir.
Benzerlik skoru: 0.4966


In [9]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-base")
llm_model = AutoModelForSeq2SeqLM.from_pretrained("google/flan-t5-base")

prompt = f"Context: {en_alakali_cumle}\nQuestion: {soru}\nAnswer:"

girisler = tokenizer(prompt, return_tensors="pt")
cikti = llm_model.generate(**girisler, max_length=50)

cevap = tokenizer.decode(cikti[0], skip_special_tokens=True)

print(f"Soru: {soru}")
print(f"Kullanılan bağlam (context): {en_alakali_cumle}")
print(f"\nLLM'in ürettiği cevap: {cevap}")

tokenizer_config.json:   0%|          | 0.00/2.54k [00:00<?, ?B/s]

spiece.model: reconstructing file:   0%|          |  0.00B /  792kB            

spiece.model: downloading bytes:           |  0.00B            

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

model.safetensors: reconstructing file:   0%|          |  0.00B /  990MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/282 [00:00<?, ?it/s]

[transformers] The tied weights mapping and config for this model specifies to tie shared.weight to lm_head.weight, but both are present in the checkpoints with different values, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning.


generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

Soru: PyTorch'u kim geliştirdi?
Kullanılan bağlam (context): Python, 1991 yılında Guido van Rossum tarafından geliştirilen, okunabilirliği yüksek bir programlama dilidir.

LLM'in ürettiği cevap: Guido van Rossum


# RAG ile LLM Tabanlı Soru-Cevap Denemesi

Bu çalışmada, bir dil modeli (LLM) yüklenerek **RAG** (Retrieval-Augmented Generation: "getirmeyle-güçlendirilmiş üretim") mantığında bir soru-cevap sistemi kuruldu. Amaç, modele bir soru sorarken yanında ilgili bir **bağlam** (context) metni vererek cevabın bu bağlamdan üretilmesini sağlamak ve bağlamın cevaba etkisini gözlemlemekti.

## Amaç

- Bir LLM'i Hugging Face üzerinden yükleyip çalıştırmak.
- Modele soru + bağlam vererek RAG mantığında cevap ürettirmek.
- Verilen bağlamın, modelin ürettiği cevabı nasıl belirlediğini gözlemlemek.

## Kullanılan Araçlar

- **Python**
- **Hugging Face Transformers** (LLM'leri yükleyip çalıştırmak için kütüphane)
- **PyTorch** (modelin arka planda çalıştığı derin öğrenme kütüphanesi)
- **Google Colab** (çalışma ortamı)

## Yapılan Adımlar

1. Modelin `safetensors` ağırlıkları (weights: modelin öğrendiği milyonlarca parametre) indirildi ve belleğe yüklendi.
2. Modelin üretim ayarları (`generation_config.json`) yüklendi.
3. Modele bir **soru** ve bu soruyla ilişkili bir **bağlam** metni verildi.
4. Modelin, cevabı kendi ezberinden değil, verilen bağlamdan üretmesi sağlandı.

## Örnek Çalışma

| Alan | İçerik |
|------|--------|
| **Soru** | PyTorch'u kim geliştirdi? |
| **Verilen Bağlam** | Python, 1991 yılında Guido van Rossum tarafından geliştirilen, okunabilir bir programlama dilidir. |
| **Modelin Cevabı** | Guido van Rossum |

## Gözlem ve Çıkarılan Ders

Modelin verdiği cevap, PyTorch açısından **yanlıştır** (PyTorch aslında Meta / Facebook AI Research ekibi tarafından geliştirilmiştir; Guido van Rossum ise Python'un geliştiricisidir).

Ancak bu bir model hatası değildir: Model, **kendisine verilen bağlama sadık kalmıştır.** Bağlam Python hakkında olduğu için, model bu metinden çekebildiği tek ismi ("Guido van Rossum") cevap olarak vermiştir.

Bu deneme, RAG sistemlerinin temel kuralını net biçimde göstermektedir:

> **Model, kendisine getirilen bağlam kadar iyidir.** Yanlış veya alakasız bir bağlam getirilirse, doğru çalışan bir model bile yanlış cevap üretir ("garbage in, garbage out").

Yani asıl belirleyici kısım modelin kendisi değil, doğru bağlamı bulup getiren **retrieval** (soruya en uygun belgeyi arayıp seçme) adımıdır.

## Sonuç

RAG mantığında bir soru-cevap kurulumu başarıyla oluşturuldu ve çalıştırıldı. Deneme, bağlam kalitesinin cevabın doğruluğu üzerindeki belirleyici etkisini somut bir örnekle ortaya koydu. Bir sonraki adımda, soruya uygun (doğru) bir bağlam verilerek modelin cevabının nasıl değiştiği test edilebilir.